storm-server/src/test/java/org/apache/storm/security/auth/AuthTest.java

```diff
@@ -0,0 +1,637 @@
+        assertFalse(authorizer.permit(notAdmin, "fileUpload", null));
+        assertFalse(authorizer.permit(userB, "fileUpload", null));
+
+        assertTrue(authorizer.permit(supervisor, "fileDownload", null));
+    }
+
+    @Test
+    public void simpleAclSameUserAuthTest() {
+        Map<String, Object> clusterConf = ConfigUtils.readStormConfig();
+        clusterConf.put(Config.NIMBUS_ADMINS, Arrays.asList("admin"));
+        clusterConf.put(Config.NIMBUS_SUPERVISOR_USERS, Arrays.asList("admin"));
+        ReqContext admin = new ReqContext(mkSubject("admin"));
+
+        final Map<String, Object> empty = Collections.emptyMap();
+        final Map<String, Object> aAllowed = new HashMap<>();
+        aAllowed.put(Config.TOPOLOGY_USERS, Arrays.asList("user-a"));
+
+        SimpleACLAuthorizer authorizer = new SimpleACLAuthorizer();
+        authorizer.prepare(clusterConf);
+
+        assertTrue(authorizer.permit(admin, "submitTopology", empty));
+        assertTrue(authorizer.permit(admin, "fileUpload", null));
+        assertTrue(authorizer.permit(admin, "getNimbusConf", null));
+        assertTrue(authorizer.permit(admin, "getClusterInfo", null));
+        assertTrue(authorizer.permit(admin, "fileDownload", null));
+        assertTrue(authorizer.permit(admin, "killTopology", aAllowed));
+        assertTrue(authorizer.permit(admin, "uploadNewCredentials", aAllowed));
+        assertTrue(authorizer.permit(admin, "rebalance", aAllowed));
+        assertTrue(authorizer.permit(admin, "activate", aAllowed));
+        assertTrue(authorizer.permit(admin, "getTopologyConf", aAllowed));
+        assertTrue(authorizer.permit(admin, "getTopology", aAllowed));
+        assertTrue(authorizer.permit(admin, "getUserTopology", aAllowed));
+        assertTrue(authorizer.permit(admin, "getTopologyInfo", aAllowed));
+    }
+
+    @Test
+    public void shellBaseGroupsMappingTest() throws Exception {
+        Map<String, Object> clusterConf = ConfigUtils.readStormConfig();
+        ShellBasedGroupsMapping groups = new ShellBasedGroupsMapping();
+        groups.prepare(clusterConf);
+
+        String userName = System.getProperty("user.name");
+
+        assertTrue(groups.getGroups(userName).size() >= 0);
+        assertEquals(0, groups.getGroups("userDoesNotExist").size());
+        assertEquals(0, groups.getGroups(null).size());
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void getTransportPluginThrowsRunimeTest() {
+        Map<String, Object> conf = ConfigUtils.readStormConfig();
+        conf.put(Config.STORM_THRIFT_TRANSPORT_PLUGIN, "null.invalid");
+        AuthUtils.getTransportPlugin(ThriftConnectionType.NIMBUS, conf, null);
+    }
+
+    public static ReqContext mkImpersonatingReqContext(String impersonatingUser, String userBeingIUmpersonated, InetAddress remoteAddress) {
+        ReqContext ret = new ReqContext(mkSubject(userBeingIUmpersonated));
+        ret.setRemoteAddress(remoteAddress);
```

Comment:
userBeingImpersonated ?

Reply:
In SASL there is the concept of a user acting on behalf of another user, or impersonating them.

We use this with the UI.

The UI user will make a connection to nimbus and authenticate as itself, but as part of the request say that it is acting on behalf of the actual user that is looking at the page.  That way nimbus can verify not only that the UI user is allowed to impersonate other users, but also that the user it is impersonating has the right permissions to do what the UI is trying to do.